### PR TITLE
A5

### PR DESCRIPTION
**Pull request by Asai-Yume**
I still have some issues which seem to be some errors with the functions in partinfo, so I do not want to close the issue just yet

### CODE REVIEW

File: stepcvt/project.py [hunk 1]
```diff
@@ -30,9 +30,33 @@ def __init__(self, name: str = "", path: Path = None, partinfo: list = None):
...
+    def to_dict(self, root=None):     
+        if isinstance(self.path, str):
+            path = Path(self.path)
+        else:
+            path = self.path
+
+        if path.is_absolute():
+            if not root:
+                raise AssertionError("Root must be provided for absolute paths!")
+            path = PurePosixPath(path).relative_to("C:" / root)
```
Comment (sree314) [inline stepcvt/project.py:42]: This assumes a Windows system. I think you can just do relative_to (root)

File: stepcvt/project.py [hunk 2]
```diff
@@ -30,9 +30,33 @@ def __init__(self, name: str = "", path: Path = None, partinfo: list = None):
...
+        else:
+            path = path
+
+        return {"type": "CADSource",
+                "name": self.name,
+                "path": str(path).replace("\\", "/"),
```
Comment (sree314) [inline stepcvt/project.py:48]: Not sure why you're doing this?

File: stepcvt/project.py [hunk 3]
```diff
@@ -30,9 +30,33 @@ def __init__(self, name: str = "", path: Path = None, partinfo: list = None):
...
+                "partinfo": self.partinfo
+                }
+    
+    def from_dict(self, root=None):       
+        if root:
+            path = Path("C:" / root / self["path"])
```
Comment (sree314) [inline stepcvt/project.py:54]: Windows-specific.

File: tests/test_cadsource.py [hunk 4]
```diff
@@ -23,23 +25,24 @@ def test_CADSource_to_dict():
 
 
 def test_CADSource_to_dict_absolute():
-    x = CADSource(name="Rapido Hotend", path=Path("/tmp/abc/xyz.step"))
+    x = CADSource(name="Rapido Hotend", path=Path("C:/tmp/abc/xyz.step"))
```
Comment (sree314) [inline tests/test_cadsource.py:28]: Don't change this.

File: tests/test_cadsource.py [hunk 5]
```diff
@@ -23,23 +25,24 @@ def test_CADSource_to_dict():
...
     assert d["path"] == "abc/xyz.step"  # note: this is not a path, but a string
 
     p = Path(d["path"])
     assert not p.is_absolute()
 
 
 def test_CADSource_to_dict_absolute_error():
-    x = CADSource(name="Rapido Hotend", path=Path("/tmp/abc/xyz.step"))
-
+    x = CADSource(name="Rapido Hotend", path=Path("C:/tmp/abc/xyz.step"))
```
Comment (sree314) [inline tests/test_cadsource.py:45]: Windows-specific change.

File: tests/test_cadsource.py [hunk 6]
```diff
@@ -60,7 +63,7 @@ def test_CADSource_from_dict():
     # the root parameter should be combined with the path stored in
     # the dictionary and converted into an absolute path when processing
     assert cs.path.is_absolute()
-    assert str(cs.path) == "/tmp/abc/xyz.step"
+    assert str(cs.path).replace("\\", "/") == "C:/tmp/abc/xyz.step" # modified to accomodate windows path
```
Comment (sree314) [inline tests/test_cadsource.py:66]: Maybe look at Pathlib to figure out if there's a way to encode portable paths?

**Review by sree314 — CHANGES_REQUESTED** [review]
I think the code has too much Windows-specific stuff like drive names.

**Comment by Asai-Yume** [comment]
> I think the code has too much Windows-specific stuff like drive names. 

If I remove the codes to accomodate the windows system, the tests wouldn't pass. Should we allow some sort of parameters to modify the paths? Or is there something I'm missing here?

**Comment by sree314** [comment]
Use the `as_posix` function for the slashes. I think removing the drive is necessary, not sure why it doesn't get removed in to_relative(). I think the tests may need some modifications, but reading Pathlib documentation may provide some better ideas that don't result in Windows- or Linux-specific code.

**Comment by Asai-Yume** [comment]
> Use the `as_posix` function for the slashes. I think removing the drive is necessary, not sure why it doesn't get removed in to_relative(). I think the tests may need some modifications, but reading Pathlib documentation may provide some better ideas that don't result in Windows- or Linux-specific code.

**Comment by Asai-Yume** [comment]
> Use the `as_posix` function for the slashes. I think removing the drive is necessary, not sure why it doesn't get removed in to_relative(). I think the tests may need some modifications, but reading Pathlib documentation may provide some better ideas that don't result in Windows- or Linux-specific code.

Yes, using the as_posix function did resolve the problems for slashes. I think what needed to be done is using a PurePath object rather than a Path object. Though I'm still unsure what to do with the from_dict function, since I'm manually padding a drive to the path for windows users, and I do not see functions in the documentation that does that.

File: stepcvt/project.py [hunk 7]
```diff
@@ -30,9 +30,30 @@ def __init__(self, name: str = "", path: Path = None, partinfo: list = None):
...
+
+        return {"type": "CADSource",
+                "name": self.name,
+                "path": str(path.as_posix()),
+                "partinfo": self.partinfo
+                }
+    
+    def from_dict(self, root=None):       
+        if root:
+            path = Path("C:" / root / self["path"])
```
Comment (sree314) [inline stepcvt/project.py:51]: I think you can remove the drive here, since it'll be present in root if needed.

File: stepcvt/project.py [hunk 8]
```diff
@@ -30,9 +30,30 @@ def __init__(self, name: str = "", path: Path = None, partinfo: list = None):
...
-
+    def to_dict(self, root=None):     
+        path = PurePath(self.path)
+        drive = path.drive
+
+        if path.is_absolute():
+            if not root:
+                raise AssertionError("Root must be provided for absolute paths!")
+            if drive:
+                path = path.relative_to(drive / root)
```
Comment (sree314) [inline stepcvt/project.py:41]: This will not produce relative paths on Unix-like systems.

**Comment by sree314** [comment]
Don't add C: -- that's wrong even on Windows when multiple drives are used.

File: tests/test_cadsource.py (hunk 5 above)
Comment (sree314) [reply tests/test_cadsource.py:45]: I've been thinking about this, and I think we need two tests -- one for Unix-like systems and another for Windows-like systems. The Windows tests can add a drive letter, etc.

File: tests/test_cadsource.py (hunk 6 above)
Comment (sree314) [reply tests/test_cadsource.py:66]: This would also go into the Windows specific test.

**Review by sree314 — APPROVED** [review]

**Pull request merged**